File: fusion_core/python_prototype/physics_speed.py
"""Channel P - physics-based forward-speed estimator.

This is **not** Channel A. Channel A (MIP Section 4.2) is a learned TCN
regressor whose slot in the UKF's sequential update order stays exactly
where it is; this module occupies the same *slot* with a classical,
untrained estimator so the filter has some velocity evidence during a
GNSS blackout instead of none at all.

Why it exists
-------------
`ukf.fx` is a CTCV process model. It holds speed magnitude constant
across the prediction step and only rotates it into the new
gyro-propagated heading. It never integrates accelerometer. Every
velocity *change* therefore has to arrive through a measurement update.
With Channel A and Channel B absent - which is the honest state of this
repo, see `models/README.md` - the filter has zero velocity evidence
during a blackout and simply coasts at whatever speed it last saw.

Channel P gives it something. The chain is:

1. `MountLeveling.from_stationary_window` estimates the gravity
   direction from a short stationary window at session start. This
   stands in for the untrained `alignment_net` (MIP Section 4.1) and
   makes no claim to be as good as one.
2. `MountLeveling.horizontal` rotates raw 3-axis accelerometer into a
   gravity-aligned level frame and subtracts gravity, leaving the
   horizontal specific force.
3. `estimate_forward_axis` resolves the remaining yaw ambiguity - which
   horizontal direction is "forward" - by PCA over a driving window,
   since accelerate/brake events dominate the horizontal acceleration
   variance and lie along the vehicle's longitudinal axis.
4. `PhysicsSpeedChannel` integrates that longitudinal component into a
   scalar forward speed, reseeding from GNSS speed whenever GNSS is
   available.

Steps 1-3 are the front-end and are only needed when the caller has raw
3-axis accelerometer in an unknown mount orientation (i.e. a phone).
Callers that already have a body-frame longitudinal component - the
benchmark tool's synthetic routes, or anything downstream of a real
alignment net - can skip straight to `PhysicsSpeedChannel` and feed it
that scalar.

Honest statement of what this is
--------------------------------
This channel drifts. It is an open-loop single integration of a biased
accelerometer, so its speed error grows roughly linearly with the time
since the last GNSS reseed, and the resulting position error grows
roughly quadratically. That is expected and is not a bug to be tuned
away. It is also why `PhysicsSpeedConfig.r_mps` must be an empirically
measured number and not an aspirational one: see
`tools/measure_physics_channel_r.py`, which measures it against the
benchmark route rather than asserting it.

One caveat worth stating plainly, because it decides whether this
channel helps at all on a given route: a velocity channel can only
improve on the CTCV coast when the vehicle's speed actually *changes*
during the blackout. On a perfectly constant-speed route the coast
model is already exactly right, and any noisy velocity channel -
including this one - can only make things worse. See
`fusion_core/python_prototype/README.md` for the measured numbers on
both a constant-speed and a speed-varying synthetic route.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np

GRAVITY_MPS2 = 9.80665


# --- front-end: mount leveling (stands in for the untrained alignment_net) ---


@dataclass
class MountLeveling:
    """Gravity-aligned rotation from the phone's (unknown) mount frame
    into a level frame whose z axis points up.

    Resolves pitch and roll only. The remaining yaw degree of freedom -
    which horizontal direction the vehicle actually faces - is *not*
    determined by gravity and is resolved separately by
    `estimate_forward_axis`.
    """

    rotation: np.ndarray  # (3, 3), body -> level
    gravity_magnitude: float

    @classmethod
    def from_stationary_window(cls, accel_xyz: np.ndarray) -> "MountLeveling":
        """Estimate leveling from a window of 3-axis accelerometer
        samples taken while the vehicle is stationary.

        `accel_xyz` is (N, 3) in m/s^2, raw mount frame. At rest the
        accelerometer measures specific force, which is the reaction to
        gravity - so its mean points *up* in the world frame, not down.
        """
        accel_xyz = np.asarray(accel_xyz, dtype=float)
        if accel_xyz.ndim != 2 or accel_xyz.shape[1] != 3:
            raise ValueError(f"expected (N, 3) accelerometer window, got {accel_xyz.shape}")
        if len(accel_xyz) < 2:
            raise ValueError("need at least 2 samples to estimate a gravity vector")

        mean = accel_xyz.mean(axis=0)
        magnitude = float(np.linalg.norm(mean))
        if magnitude < 1.0:
            raise ValueError(
                f"stationary-window mean |accel| is {magnitude:.3f} m/s^2, which is "
                "not a plausible gravity magnitude - the window is probably not "
                "stationary, or the accelerometer is already gravity-compensated "
                "(in which case skip MountLeveling entirely)."
            )

        up = mean / magnitude
        # Gram-Schmidt an arbitrary body axis against `up` to get a
        # level-frame x. Which horizontal direction this lands on is
        # arbitrary and is exactly what estimate_forward_axis resolves.
        seed = np.array([1.0, 0.0, 0.0])
        if abs(float(np.dot(seed, up))) > 0.9:
            seed = np.array([0.0, 1.0, 0.0])
        x_level = seed - np.dot(seed, up) * up
        x_level /= np.linalg.norm(x_level)
        y_level = np.cross(up, x_level)

        rotation = np.stack([x_level, y_level, up], axis=0)
        return cls(rotation=rotation, gravity_magnitude=magnitude)

    def horizontal(self, accel_xyz: np.ndarray) -> np.ndarray:
        """Rotate raw accelerometer into the level frame, remove
        gravity, and return the horizontal (x, y) components.

        Accepts a single (3,) sample or an (N, 3) window; returns (2,)
        or (N, 2) correspondingly.
        """
        accel_xyz = np.asarray(accel_xyz, dtype=float)
        single = accel_xyz.ndim == 1
        if single:
            accel_xyz = accel_xyz[None, :]
        leveled = accel_xyz @ self.rotation.T
        # The level-frame z component is gravity plus any genuine
        # vertical motion; horizontal is untouched by the subtraction,
        # so we simply drop z rather than subtracting from x/y.
        horizontal = leveled[:, :2]
        return horizontal[0] if single else horizontal


def estimate_forward_axis(
    horizontal_accel: np.ndarray,
    reference_speed_delta: np.ndarray | None = None,
) -> np.ndarray:
    """Resolve which horizontal direction is vehicle-forward.

    `horizontal_accel` is (N, 2) level-frame horizontal specific force over a
    window of *driving* (not stationary).

    Two estimators, and which one runs depends on whether a speed reference is
    available:

    **With `reference_speed_delta`** (N,), typically GNSS speed differences over
    the same window: the axis is the direction whose acceleration actually
    explains the speed change, computed as the cross-correlation vector
    `centered.T @ delta`, normalised. This resolves the axis and its sign
    together, in one step.

    **Without it**: first principal component, on the assumption that
    accelerate and brake events dominate horizontal variance. PCA gives an axis,
    not a direction, so the sign is left arbitrary and the caller owns the
    ambiguity.

    Why the correlation estimator is preferred, measured rather than asserted
    ------------------------------------------------------------------------
    PCA's assumption fails on any window where turning dominates braking. In a
    turn, lateral specific force is `speed * yaw_rate`; at 13 m/s and 6 deg/s
    that is about 1.4 m/s^2, while the longitudinal content of normal
    speed variation on the same route is around 0.5 m/s^2. So a window with a
    high enough proportion of turning has its variance dominated by the
    *lateral* axis, and PCA confidently returns the axis at ninety degrees to
    the right answer. Channel P then integrates cornering force as though it
    were acceleration.

    This is not hypothetical. Measured against the known mount rotation of
    `tools/phone_replay/synth_session.py`, as dot product with the true forward
    axis, over the first N moving samples:

    | samples | PCA    | correlation |
    |---------|--------|-------------|
    | 1000    | 1.0000 | 1.0000      |
    | 3000    | 0.9747 | 0.9839      |
    | 5000    | 0.9686 | 0.9824      |
    | 6000    | 0.3640 | 0.9912      |
    | 9000    | 0.9999 | 0.9981      |
    | 12000   | 1.0000 | 0.9999      |
    | 15404   | 1.0000 | 0.9999      |

    The 6000 row is a 69-degree error. PCA recovers by 9000 samples because the
    session's later straight sections dilute the turns, which is exactly why the
    offline whole-session caller never saw this: it always has 15000 samples. An
    *online* caller on a phone only has the samples up to now, and passes
    through that 6000-sample window during the first minute of every drive,
    which on a demo route is when the blackout happens.

    Turning does not change speed, so it contributes nothing to the correlation
    in expectation. That is the whole reason the second estimator is steadier,
    and it is a property of the physics rather than of this dataset.

    The trade, stated rather than hidden: given the whole session, PCA is very
    slightly the better of the two (1.0000 against 0.9999), and switching the
    offline tool to the correlation estimator moved its Channel P drift on the
    60-90 s window from 0.46% to 1.50%. That is the price. It buys a worst case
    across every window size of 0.982 instead of 0.364, and the phone is causal
    and has no choice but to run on a partial window. Both numbers are far under
    PS 26168's 10% bar, so paying a point of drift for an estimator that cannot
    silently point sideways is the right way round.

    The correlation vector degenerates when speed is genuinely constant across
    the window, since then there is no speed change to correlate against. That
    case falls back to PCA with a sign fixed the old way, which is no worse than
    what was here before.
    """
    horizontal_accel = np.asarray(horizontal_accel, dtype=float)
    if horizontal_accel.ndim != 2 or horizontal_accel.shape[1] != 2:
        raise ValueError(f"expected (N, 2) horizontal accel, got {horizontal_accel.shape}")
    if len(horizontal_accel) < 3:
        raise ValueError("need at least 3 samples to estimate a forward axis")

    centered = horizontal_accel - horizontal_accel.mean(axis=0)

    if reference_speed_delta is not None:
        reference_speed_delta = np.asarray(reference_speed_delta, dtype=float)
        n = min(len(centered), len(reference_speed_delta))
        correlation_vector = centered[:n].T @ reference_speed_delta[:n]
        magnitude = float(np.linalg.norm(correlation_vector))
        # A scale-free floor: the correlation is meaningful only when it is
        # large relative to the acceleration and speed-change magnitudes that
        # produced it. Near zero means the window carried no usable
        # longitudinal signal.
        scale = float(
            np.linalg.norm(centered[:n]) * np.linalg.norm(reference_speed_delta[:n])
        )
        if magnitude > 1e-9 and magnitude > 1e-6 * max(scale, 1e-12):
            return correlation_vector / magnitude

    _, _, vt = np.linalg.svd(centered, full_matrices=False)
    axis = vt[0]
    axis = axis / np.linalg.norm(axis)

    if reference_speed_delta is not None:
        projected = centered @ axis
        n = min(len(projected), len(reference_speed_delta))
        correlation = float(np.dot(projected[:n], reference_speed_delta[:n]))
        if correlation < 0:
            axis = -axis

    return axis


# --- the channel itself -------------------------------------------------------


@dataclass
class PhysicsSpeedConfig:
    """Tunables for Channel P.

    `r_mps` is the 1-sigma this channel should be handed to the UKF as.
    The default here is deliberately *not* a target or an aspiration -
    it is a placeholder that callers are expected to override with a
    measured number from `tools/measure_physics_channel_r.py`. Note
    also that this channel's error is a slow drift, not white noise, so
    the Kalman independence assumption behind R is violated; a measured
    RMSE is a pragmatic proxy, and a defensible one, but it is a proxy.
    """

    r_mps: float = 2.0
    max_speed_mps: float = 70.0
    # Optional first-order leak of the integrated speed toward the last
    # GNSS-confirmed speed, as a time constant in seconds. 0.0 disables
    # it. This is a bound on how far an integrated bias can run away
    # during a long blackout, at the cost of being wrong during genuine
    # sustained acceleration. Off by default - turn it on only with a
    # measured justification.
    leak_tau_s: float = 0.0


class PhysicsSpeedChannel:
    """Integrates longitudinal acceleration into a scalar forward
    speed, reseeded from GNSS whenever GNSS is available.

    Stateful across a route; construct one per run and call `update`
    once per cycle. Returns `None` until the first GNSS reseed, because
    before that the channel has no absolute reference and must not feed
    the filter a fabricated one.
    """

    def __init__(self, config: PhysicsSpeedConfig | None = None) -> None:
        self.config = config or PhysicsSpeedConfig()
        self._speed: float | None = None
        self._last_gnss_speed: float | None = None

    @property
    def speed(self) -> float | None:
        return self._speed

    def reset(self) -> None:
        self._speed = None
        self._last_gnss_speed = None

    def reseed(self, gnss_speed: float) -> float:
        """Snap the integrated speed to a GNSS speed measurement. This
        is what keeps the integration from being open-loop over the
        whole route rather than just over each blackout."""
        self._speed = float(max(0.0, gnss_speed))
        self._last_gnss_speed = self._speed
        return self._speed

    def apply_zupt(self) -> float:
        """Force the integrated speed to zero. Called when an external
        zero-velocity detector says the vehicle is stopped. Kept
        separate from `update` so the detector's decision and this
        channel's integration stay independently testable."""
        self._speed = 0.0
        return self._speed

    def update(
        self,
        dt: float,
        longitudinal_accel: float,
        gnss_speed: float | None = None,
    ) -> float | None:
        """Advance one cycle.

        `longitudinal_accel` is the forward-axis specific force in
        m/s^2, already leveled and gravity-free. `gnss_speed` is the
        GNSS speed-over-ground in m/s when available, else None.
        """
        if gnss_speed is not None:
            return self.reseed(gnss_speed)

        if self._speed is None:
            # No absolute reference yet. Returning None is the point:
            # the caller must skip the update rather than feed the
            # filter a number this channel cannot justify.
            return None

        speed = self._speed + float(longitudinal_accel) * dt

        tau = self.config.leak_tau_s
        if tau > 0.0 and self._last_gnss_speed is not None:
            alpha = min(1.0, dt / tau)
            speed = speed + alpha * (self._last_gnss_speed - speed)

        self._speed = float(np.clip(speed, 0.0, self.config.max_speed_mps))
        return self._speed

File: android_app/app/src/main/java/org/sih26/deadreckoning/fusion/PhysicsSpeed.kt
package org.sih26.deadreckoning.fusion

import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

/**
 * Kotlin port of fusion_core/python_prototype/physics_speed.py, Channel P.
 *
 * This is not Channel A. Channel A is a learned TCN regressor that was trained,
 * failed its pre-registered acceptance gate on bias, and was not shipped. Channel P
 * occupies the same slot in the UKF's update order with a classical, untrained
 * estimator, so the filter has some velocity evidence during a blackout instead of
 * none at all. Say that plainly rather than letting anyone infer a model is running.
 *
 * What this channel is, stated honestly: an open-loop single integration of a biased
 * accelerometer, reseeded from GNSS whenever GNSS is available. Its speed error grows
 * roughly linearly with time since the last reseed and the resulting position error
 * roughly quadratically. That is expected and is not a bug to be tuned away. It is
 * also why r_mps must be a measured number rather than an aspirational one.
 *
 * The three front-end pieces here exist specifically for the phone case, where the
 * accelerometer arrives in three axes pointing wherever the holder happens to point,
 * with gravity in them:
 *   1. MountLeveling resolves pitch and roll from gravity.
 *   2. ForwardAxisEstimator resolves the remaining yaw ambiguity by PCA over driving.
 *   3. PhysicsSpeedChannel integrates the longitudinal component.
 *
 * Steps 1 and 2 stand in for the untrained alignment_net and make no claim to be as
 * good as one.
 */

const val GRAVITY_MPS2 = 9.80665

/**
 * Gravity-aligned rotation from the phone's unknown mount frame into a level frame
 * whose z axis points up. Resolves pitch and roll only; which horizontal direction
 * the vehicle faces is not determined by gravity and is [ForwardAxisEstimator]'s job.
 *
 * `rotation` is stored row-major as body to level, matching the Python reference's
 * `np.stack([x_level, y_level, up], axis=0)`, so row 2 is the up vector. Yaw rate is
 * the gyro projected onto that row, never the phone's own z axis.
 */
class MountLeveling(val rotation: Array<DoubleArray>, val gravityMagnitude: Double) {

    val up: DoubleArray get() = rotation[2]

    /** Rotate a raw accelerometer sample into the level frame and return its
     * horizontal (x, y) components. Gravity lands entirely in level-frame z, so it is
     * dropped rather than subtracted from x and y. */
    fun horizontal(accel: DoubleArray): DoubleArray {
        val x = rotation[0][0] * accel[0] + rotation[0][1] * accel[1] + rotation[0][2] * accel[2]
        val y = rotation[1][0] * accel[0] + rotation[1][1] * accel[1] + rotation[1][2] * accel[2]
        return doubleArrayOf(x, y)
    }

    /** Yaw rate about the gravity-aligned vertical axis.
     *
     * Not gyro.z. The mount orientation is arbitrary, so the phone's own z axis is
     * not the vehicle's vertical, and using it is the single easiest way to make a
     * phone-mounted filter turn the wrong way. Validate the sign on a known left turn
     * before trusting any recording: if the phone is face down it flips. */
    fun yawRate(gyro: DoubleArray): Double = LinAlg.dot(gyro, up)

    companion object {
        /**
         * Estimate leveling from a window of accelerometer samples taken while the
         * vehicle is stationary. At rest the accelerometer measures specific force,
         * the reaction to gravity, so its mean points UP in the world frame, not down.
         */
        fun fromStationaryWindow(window: List<DoubleArray>): MountLeveling {
            require(window.size >= 2) { "need at least 2 samples to estimate a gravity vector" }

            val mean = DoubleArray(3)
            for (sample in window) for (i in 0 until 3) mean[i] += sample[i]
            for (i in 0 until 3) mean[i] /= window.size

            val magnitude = LinAlg.norm(mean)
            if (magnitude < 1.0) {
                throw IllegalArgumentException(
                    "stationary-window mean |accel| is $magnitude m/s^2, which is not a " +
                        "plausible gravity magnitude. Either the window was not stationary, " +
                        "or TYPE_LINEAR_ACCELERATION was used instead of TYPE_ACCELEROMETER " +
                        "and gravity has already been removed by a vendor filter."
                )
            }

            val up = DoubleArray(3) { mean[it] / magnitude }

            // Gram-Schmidt an arbitrary body axis against up to get a level-frame x.
            // Which horizontal direction this lands on is arbitrary and is exactly
            // what the forward-axis estimator resolves.
            var seed = doubleArrayOf(1.0, 0.0, 0.0)
            if (abs(LinAlg.dot(seed, up)) > 0.9) seed = doubleArrayOf(0.0, 1.0, 0.0)

            val projection = LinAlg.dot(seed, up)
            val xLevel = DoubleArray(3) { seed[it] - projection * up[it] }
            val xNorm = LinAlg.norm(xLevel)
            for (i in 0 until 3) xLevel[i] /= xNorm

            val yLevel = doubleArrayOf(
                up[1] * xLevel[2] - up[2] * xLevel[1],
                up[2] * xLevel[0] - up[0] * xLevel[2],
                up[0] * xLevel[1] - up[1] * xLevel[0]
            )

            return MountLeveling(arrayOf(xLevel, yLevel, up), magnitude)
        }
    }
}

/**
 * Resolves which horizontal direction is vehicle-forward, online.
 *
 * The estimator is the one described in the Python reference's
 * `estimate_forward_axis`: the axis is the direction whose acceleration explains
 * the GNSS speed change, computed as the cross-correlation vector between centred
 * horizontal specific force and speed delta. That resolves the axis and its sign
 * together. An unresolved or flipped sign makes Channel P integrate backwards, which
 * is worse than not having the channel at all.
 *
 * Do not "simplify" this back to a first principal component. PCA was what this used
 * to be, and it fails on exactly the windows a phone runs in. In a turn, lateral
 * specific force is speed times yaw rate, several times larger than the longitudinal
 * content of ordinary speed variation, so a turn-heavy window has its variance
 * dominated by the lateral axis and PCA returns an axis ninety degrees off. Measured
 * against a known mount rotation, PCA's worst case across window sizes was a dot
 * product of 0.364 with the true axis where this estimator's was 0.982. The offline
 * caller never saw it because a whole session dilutes the turns; an online caller
 * passes straight through that regime in the first minute of every drive.
 *
 * Accumulates running moments rather than storing samples, so memory and per-sample
 * cost are constant no matter how long the drive, and the estimate uses every moving
 * sample seen so far rather than an arbitrary recent window.
 */
class ForwardAxisEstimator(
    private val minSpeedMps: Double = 2.0,
    private val minSamples: Int = 300
) {
    private companion object {
        /** Longest gap between fixes whose speed change can still be attributed to
         * the samples inside it. Comfortably above a 1 Hz cadence with a few dropped
         * fixes, well below any deliberate blackout. */
        const val MAX_INTERVAL_S = 3.0
    }

    // Global running moments over every admitted sample.
    private var n: Long = 0
    private var sumX = 0.0
    private var sumY = 0.0
    private var sumXX = 0.0
    private var sumXY = 0.0
    private var sumYY = 0.0
    private var sumD = 0.0
    private var sumXD = 0.0
    private var sumYD = 0.0

    // Accumulators for the samples since the last GNSS fix. GNSS speed change is
    // only known once the fix that ends an interval arrives, so samples are held
    // here and folded in with the delta that actually covers them. Applying each
    // delta to the samples that follow it instead costs a full second of lag
    // between an acceleration and the speed change it caused, which decorrelates
    // the very signal this estimator runs on. Measured: with the lag, the axis
    // landed 7.4 degrees off the known mount rotation and leaked enough cornering
    // force into Channel P to push blackout drift to 38%. Without it, under a
    // degree.
    private var intervalN: Long = 0
    private var intervalSumX = 0.0
    private var intervalSumY = 0.0
    private var intervalSumXX = 0.0
    private var intervalSumXY = 0.0
    private var intervalSumYY = 0.0

    var axis: DoubleArray? = null
        private set
    var signResolved: Boolean = false
        private set

    val sampleCount: Long get() = n

    /** Buffer one sample. Nothing is committed until the interval closes. */
    fun addSample(horizontal: DoubleArray, speedMps: Double) {
        if (speedMps <= minSpeedMps) return
        val x = horizontal[0]
        val y = horizontal[1]
        intervalN++
        intervalSumX += x
        intervalSumY += y
        intervalSumXX += x * x
        intervalSumXY += x * y
        intervalSumYY += y * y
    }

    /**
     * Close the current interval with the GNSS speed change that occurred across it,
     * folding its buffered samples into the running moments.
     *
     * Since the delta is constant across the interval, the per-sample sum of
     * x_i * d reduces to d times the interval's sum of x, so no sample history is
     * needed to get the correlation exactly right.
     *
     * `intervalSeconds` guards against attributing a long gap's total speed change to
     * every sample inside it. A blackout produces exactly that gap: no fix reaches
     * the pipeline for the whole window, and folding thirty seconds of buffered
     * samples in with a thirty-second speed delta swamps every honest interval before
     * it. Measured: that single mistake moved the axis from under a degree off the
     * known mount rotation to thirteen degrees off, and took blackout drift from
     * 1.6% to 30%. Over-long intervals are discarded, not clamped: there is no
     * correct weight for them.
     */
    fun closeInterval(speedChange: Double, intervalSeconds: Double = 1.0) {
        if (intervalN == 0L) return
        if (intervalSeconds > MAX_INTERVAL_S) {
            discardInterval()
            return
        }
        n += intervalN
        sumX += intervalSumX
        sumY += intervalSumY
        sumXX += intervalSumXX
        sumXY += intervalSumXY
        sumYY += intervalSumYY
        sumD += speedChange * intervalN
        sumXD += intervalSumX * speedChange
        sumYD += intervalSumY * speedChange

        discardInterval()
    }

    /** Throw away the buffered samples without committing them. Called when the
     * speed reference that would have explained them is unavailable, which is what a
     * GNSS blackout means. */
    fun discardInterval() {
        intervalN = 0
        intervalSumX = 0.0
        intervalSumY = 0.0
        intervalSumXX = 0.0
        intervalSumXY = 0.0
        intervalSumYY = 0.0
    }

    fun hasEnough(): Boolean = n >= minSamples

    /**
     * Compute the axis from everything committed so far, or null if there is not
     * enough driving yet. The caller must treat null as "Channel P is not ready",
     * never as "assume forward is x".
     */
    fun estimate(): DoubleArray? {
        if (n < 3) return null

        val count = n.toDouble()
        val meanX = sumX / count
        val meanY = sumY / count

        // Cross-correlation of centred horizontal accel with speed change. Expanding
        // sum((x - mx) * d) gives sumXD - mx * sumD, so no sample history is needed.
        var ax = sumXD - meanX * sumD
        var ay = sumYD - meanY * sumD
        var magnitude = sqrt(ax * ax + ay * ay)

        if (magnitude > 1e-9) {
            ax /= magnitude
            ay /= magnitude
            val resolved = doubleArrayOf(ax, ay)
            axis = resolved
            signResolved = true
            return resolved
        }

        // Degenerate: speed was effectively constant across everything seen, so
        // there is nothing to correlate against. Fall back to the principal axis of
        // the centred data, with the sign left unresolved, matching the Python
        // reference's no-reference path.
        val sxx = sumXX - count * meanX * meanX
        val sxy = sumXY - count * meanX * meanY
        val syy = sumYY - count * meanY * meanY

        val trace = sxx + syy
        val determinant = sxx * syy - sxy * sxy
        val discriminant = max(0.0, trace * trace / 4.0 - determinant)
        val largest = trace / 2.0 + sqrt(discriminant)

        if (abs(sxy) > 1e-12) {
            ax = largest - syy
            ay = sxy
        } else {
            if (sxx >= syy) { ax = 1.0; ay = 0.0 } else { ax = 0.0; ay = 1.0 }
        }
        magnitude = sqrt(ax * ax + ay * ay)
        if (magnitude < 1e-12) return null

        val resolved = doubleArrayOf(ax / magnitude, ay / magnitude)
        axis = resolved
        signResolved = false
        return resolved
    }

    fun reset() {
        n = 0
        sumX = 0.0; sumY = 0.0
        sumXX = 0.0; sumXY = 0.0; sumYY = 0.0
        sumD = 0.0; sumXD = 0.0; sumYD = 0.0
        intervalN = 0
        intervalSumX = 0.0; intervalSumY = 0.0
        intervalSumXX = 0.0; intervalSumXY = 0.0; intervalSumYY = 0.0
        axis = null
        signResolved = false
    }
}


data class PhysicsSpeedConfig(
    /** The 1-sigma this channel is handed to the UKF as. Override with a measured
     * number from tools/measure_physics_channel_r.py. Note that this channel's error
     * is a slow drift rather than white noise, so the independence assumption behind
     * R is violated; a measured RMSE is a pragmatic proxy, and a defensible one, but
     * it is a proxy. */
    val rMps: Double = 2.0,
    val maxSpeedMps: Double = 70.0,
    /** Optional first-order leak of integrated speed toward the last GNSS-confirmed
     * speed, as a time constant in seconds. Zero disables it. It bounds how far an
     * integrated bias can run away during a long blackout, at the cost of being wrong
     * during genuine sustained acceleration. Off by default; turn it on only with a
     * measured justification. */
    val leakTauS: Double = 0.0
)

/**
 * Integrates longitudinal acceleration into a scalar forward speed, reseeded from
 * GNSS whenever GNSS is available.
 *
 * Returns null until the first reseed, because before that the channel has no
 * absolute reference and must not feed the filter a fabricated one. The caller skips
 * the update on null rather than substituting zero.
 */
class PhysicsSpeedChannel(val config: PhysicsSpeedConfig = PhysicsSpeedConfig()) {

    var speed: Double? = null
        private set
    private var lastGnssSpeed: Double? = null

    fun reset() {
        speed = null
        lastGnssSpeed = null
    }

    /** Snap the integrated speed to a GNSS measurement. This is what keeps the
     * integration from being open-loop over the whole route rather than just over
     * each blackout. */
    fun reseed(gnssSpeed: Double): Double {
        val seeded = max(0.0, gnssSpeed)
        speed = seeded
        lastGnssSpeed = seeded
        return seeded
    }

    /** Force the integrated speed to zero. Called when an external detector says the
     * vehicle is stopped. Kept separate from [update] so the detector's decision and
     * this channel's integration stay independently testable. */
    fun applyZupt(): Double {
        speed = 0.0
        return 0.0
    }

    fun update(dt: Double, longitudinalAccel: Double, gnssSpeed: Double? = null): Double? {
        if (gnssSpeed != null) return reseed(gnssSpeed)

        val current = speed ?: return null

        var next = current + longitudinalAccel * dt

        val tau = config.leakTauS
        val anchor = lastGnssSpeed
        if (tau > 0.0 && anchor != null) {
            val alpha = min(1.0, dt / tau)
            next += alpha * (anchor - next)
        }

        val clamped = min(max(next, 0.0), config.maxSpeedMps)
        speed = clamped
        return clamped
    }
}
